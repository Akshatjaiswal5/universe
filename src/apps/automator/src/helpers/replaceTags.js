// @flow

import os from 'os';

export default function replaceTags(
  originalText: string,
  taskIdentifier: string,
  replacement: string,
) {
  const OPEN_TAG = `<!-- AUTOMATOR:${taskIdentifier} -->` + os.EOL;
  const CLOSE_TAG = os.EOL + `<!-- /AUTOMATOR:${taskIdentifier} -->`;
  const TAGS_REGEXP = new RegExp(
    '<!-- AUTOMATOR:' +
      taskIdentifier +
      ' -->[\\s\\S]*?<!-- /AUTOMATOR:' +
      taskIdentifier +
      ' -->',
    'gu',
  );

  return originalText.replace(TAGS_REGEXP, OPEN_TAG + replacement + CLOSE_TAG);
}
