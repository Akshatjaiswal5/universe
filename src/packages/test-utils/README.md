TODO

# `generateTestsFromFixtures`

```js
import { generateTestsFromFixtures } from '@kiwicom/test-utils';

function operation(input) {
  return doSomethingWithTheInput(input);
}

generateTestsFromFixtures(`${__dirname}/__fixtures__`, operation);
```
