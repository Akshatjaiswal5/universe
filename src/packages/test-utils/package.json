{
  "name": "@kiwicom/test-utils",
  "description": "Set of tools for easier testing in JS.",
  "homepage": "https://gitlab.skypicker.com/graphql/graphql/tree/master/src/packages/test-utils",
  "private": false,
  "version": "0.3.0",
  "main": "src/index.js",
  "dependencies": {
    "@mrtnzlml/utils": "^1.1.1"
  }
}
