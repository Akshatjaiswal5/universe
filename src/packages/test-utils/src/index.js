// @flow

import { invariant } from '@mrtnzlml/utils';

const FIXTURE_TAG = Symbol.for('FIXTURE_TAG');

/**
 * Extend Jest with a custom snapshot serializer to provide additional context
 * and reduce the amount of escaping that occurs. This serializer is heavily
 * inspired by Relay test cases.
 */
expect.addSnapshotSerializer({
  print(value) {
    return Object.keys(value)
      .map(key => `~~~~~~~~~~ ${key.toUpperCase()} ~~~~~~~~~~\n${value[key]}`)
      .join('\n');
  },
  test(value) {
    return value && value[FIXTURE_TAG] === true;
  },
});

/**
 * Generates a set of jest snapshot tests that compare the output of the
 * provided `operation` to each of the matching files in the `fixturesPath`.
 */
export function generateTestsFromFixtures(
  fixturesPath: string,
  operation: (input: string) => string | Promise<string>,
): void {
  const fs = require('fs');
  const path = require('path');
  const tests = fs.readdirSync(fixturesPath).map(async file => {
    const input = fs.readFileSync(path.join(fixturesPath, file), 'utf8');
    const output = await getOutputForFixture(input, operation);
    return {
      file,
      input,
      output,
    };
  });
  invariant(
    tests.length > 0,
    'generateTestsFromFixtures: No fixtures found at %s',
    fixturesPath,
  );
  it('matches expected output', async () => {
    const results = await Promise.all(tests);
    results.forEach(test => {
      expect({
        // $FlowIssue: https://github.com/facebook/flow/issues/3258
        [FIXTURE_TAG]: true,
        input: test.input,
        output: test.output,
      }).toMatchSnapshot(test.file);
    });
  });
}

async function getOutputForFixture(
  input: string,
  operation: (input: string) => string | Promise<string>,
): Promise<string> {
  try {
    const output = operation(input);
    return output instanceof Promise ? await output : output;
  } catch (e) {
    return `THROWN EXCEPTION:\n\n${e.toString()}`;
  }
}
