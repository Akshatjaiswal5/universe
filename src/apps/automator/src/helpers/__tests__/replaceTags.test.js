// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import replaceTags from '../replaceTags';

function operation(taskIdentifier) {
  return input => replaceTags(input, taskIdentifier, 'replaced content ðŸŽ‰');
}

describe('correct task identifier should replace the tags', () => {
  generateTestsFromFixtures(
    `${__dirname}/__fixtures__`,
    operation('TEST_FIXTURE_1'),
  );
});

describe('unknown task identifier should return unchanged input', () => {
  generateTestsFromFixtures(
    `${__dirname}/__fixtures__`,
    operation('UNKNOWN_TASK_IDENTIFIER'),
  );
});
