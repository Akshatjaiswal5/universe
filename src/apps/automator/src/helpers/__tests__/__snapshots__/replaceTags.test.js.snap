// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct task identifier should replace the tags matches expected output: emptyTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 --><!-- /AUTOMATOR:TEST_FIXTURE_1 -->

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replaced content ðŸŽ‰
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

`;

exports[`correct task identifier should replace the tags matches expected output: missingTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
just a regular text ðŸ¤“

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
just a regular text ðŸ¤“

`;

exports[`correct task identifier should replace the tags matches expected output: multipleTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replace me please
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

<!-- AUTOMATOR:TEST_FIXTURE_1 -->
and me too!
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replaced content ðŸŽ‰
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replaced content ðŸŽ‰
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

`;

exports[`correct task identifier should replace the tags matches expected output: nonEmptyTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
a <!-- AUTOMATOR:TEST_FIXTURE_1 --> b <!-- /AUTOMATOR:TEST_FIXTURE_1 --> c

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
a <!-- AUTOMATOR:TEST_FIXTURE_1 -->
replaced content ðŸŽ‰
<!-- /AUTOMATOR:TEST_FIXTURE_1 --> c

`;

exports[`unknown task identifier should return unchanged input matches expected output: emptyTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 --><!-- /AUTOMATOR:TEST_FIXTURE_1 -->

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 --><!-- /AUTOMATOR:TEST_FIXTURE_1 -->

`;

exports[`unknown task identifier should return unchanged input matches expected output: missingTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
just a regular text ðŸ¤“

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
just a regular text ðŸ¤“

`;

exports[`unknown task identifier should return unchanged input matches expected output: multipleTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replace me please
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

<!-- AUTOMATOR:TEST_FIXTURE_1 -->
and me too!
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
<!-- AUTOMATOR:TEST_FIXTURE_1 -->
replace me please
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

<!-- AUTOMATOR:TEST_FIXTURE_1 -->
and me too!
<!-- /AUTOMATOR:TEST_FIXTURE_1 -->

`;

exports[`unknown task identifier should return unchanged input matches expected output: nonEmptyTags.md 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
a <!-- AUTOMATOR:TEST_FIXTURE_1 --> b <!-- /AUTOMATOR:TEST_FIXTURE_1 --> c

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
a <!-- AUTOMATOR:TEST_FIXTURE_1 --> b <!-- /AUTOMATOR:TEST_FIXTURE_1 --> c

`;
